drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+      target_audio = @target_audio
+      if target_audio && self[:audio] != self[:target_audio]
+        switch_audio(target_audio)
+      end
+    when .volume_status?
+      self[:volume_max] = max
+      unless self[:audio_mute]
+        self[:volume] = value
+      end
+    when .brightness_status?
+      self[:brightness_max] = max
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast_max] = max
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      if(value == 1)
+        self[:volume] = 0
+      else
+        volume_status(60) # high priority
+      end
+    when .power_on_delay?
+      if value > 0
+        self[:warming] = true
+        schedule.in(value.seconds) do # Prevent any commands being sent until the power on delay is complete
+          power_on_delay.as(PlaceOS::Driver::Transport)
+        end
+      else
+        schedule.in(3.seconds) do # Reactive the interface once the display is online
+          self[:warming] = false # allow access to the display
+        end
+      end
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType
+    Command             = 0x41 # 'A'
+    Command_reply       = 0x42 # 'B'
+    Get_parameter       = 0x43 # 'C'
+    Get_parameter_reply = 0x44 # 'D'
+    Set_parameter       = 0x45 # 'E'
+    Set_parameter_reply = 0x46 # 'F'
+  end
+
+  enum Command
+    Video_input       = 0x0060
+    Audio_input       = 0x022E
+    Volume_status     = 0x0062
+    Mute_status       = 0x008D
+    Power_on_delay    = 0x02D8
+    Contrast_status   = 0x0012
+    Brightness_status = 0x0010
+    Auto_setup        = 0x001E
+    Power_query       = 0x01D6
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+
+  def self.format_value(value : Int, length : Int = 4) : String
+    value.to_s(16).upcase.rjust(length, '0')
+  end
+
+  {% for name in Command.constants.map(&.downcase) %}
+  @[Security(Level::Administrator)]
+    def {{name.id}}(priority : Int32 = 0)
+      do_send(MsgType::Get_parameter, Command.parse({{name}}), priority: priority, name: {{name.id}})
+    end
+  {% end %}
+
+  private def check_checksum(data : Bytes)
+    checksum = 0x00_u8
+    # Loop through the second to the third last element
+    if data.size >= 2
+      data[1..-3].each do |b|
+        checksum = checksum ^ b
+      end
+      # Check the checksum equals the second last element
+      checksum == data[-2]
+    else
+      true
+    end
+  end
+
+  # Builds the command and creates the checksum
+  # data can be an ascii encoded string
+  private def do_send(type : MsgType, data : Command | String, **options)
```

Comment:
Rather than passing `data` in as a string, I'd recommend keeping everything as enum's, or byte arrays until they reach here. That way you can perform all the mapping needed to convert to the hex string in only place.

Reply:
I don't think it's possible to make `data` only take in an enum especially for the input switching commands as they both need a command enum and the actual input enum like here

https://github.com/PlaceOS/drivers/blob/e7f3d475fd30365e5e91ac93b94ed0a1081cabf3/drivers/nec/display/display.cr#L110

which data would evaluate to `"00600001"` if `input == Input::Vga`. I can convert `data` to bytes before giving it to `do_send` but seems easier to allow data to also take a string. Let me know if I'm going about this the wrong way or if there's a better way of doing this.

Reply:
See [other comments](https://github.com/PlaceOS/drivers/pull/50#discussion_r492443003) about reducing the number different representation formats that you are dealing with at any time.

The transport for this protocol sucks, but you have two clean delineation points where you can limit the impact of this oddity:
1. `do_send`
2. `received`

If you keep all mapping to/from the ascii encoded hex string in these places, all other sections of your code can deal with the protocol in it's direct format, without worrying about conversion, or mixing converted and non converted components.

Reply:
I've given this a try but I don't think this is easy to achieve because of the way this protocol is implemented. There is no arbitrary way to know how long data needs to be or how many zeros need to be needed. For example, Command's need to of length 4 which is fine but Inputs need to be of length 2 as well as the state for power if we are setting it ("0004" for on and "0001" for off). Same thing for setting brightness and volume which will be 0x64 or 100 at most but is required to be of length 4.

Oh another case that made this more of a pain is that the Command to Save is required to be of length 2 like this `    do_send(MsgType::Command, "0C", name: "brightness_save")`

Reply:
I've made this commit https://github.com/PlaceOS/drivers/pull/50/commits/d48f502099ee72ee12c300e562fbeedf1151df4f but I'm not sure why I'm getting this error

Reply:
-> spec calling: power {false}
level=ERROR time=2020-10-05T15:48:28Z source=spec_runner message=executing {"__exec__":"power","power":[false]} on Nec::Display::All (spec_runner)
Expected Int but was Bool at 1:1 (JSON::ParseException)
  from ../usr/share/crystal/src/json/pull_parser.cr:528:5 in 'parse_exception'
  from ../usr/share/crystal/src/json/pull_parser.cr:520:5 in 'expect_kind'
  from ../usr/share/crystal/src/json/pull_parser.cr:126:5 in 'read_int'
  from ../usr/share/crystal/src/json/from_json.cr:76:1 in 'new'
  from ../usr/share/crystal/src/json/from_json.cr:13:3 in 'from_json'
  from drivers/nec/display/display.cr:8:1 in '->'
  from ../usr/share/crystal/src/primitives.cr:255:3 in 'execute'
  from lib/placeos-driver/src/placeos-driver/driver_manager.cr:128:5 in 'execute'
  from lib/placeos-driver/src/placeos-driver/driver_manager.cr:150:18 in 'process'
  from lib/placeos-driver/src/placeos-driver/driver_manager.cr:138:9 in '->'
  from ../usr/share/crystal/src/primitives.cr:255:3 in 'run'
  from ../usr/share/crystal/src/fiber.cr:92:34 in '->'
  from ???

level=ERROR : timeout waiting for expected data
-> expecting: "\u00010*0A0C\u0002C203D60004\u0003\u001D\r"
level=ERROR : unhandled exception in spec
Channel is closed (Channel::ClosedError)
  from ../usr/share/crystal/src/channel.cr:273:20 in 'receive'
  from lib/placeos-driver/src/placeos-driver/driver-specs/runner.cr:433:16 in 'should_send'
  from lib/placeos-driver/src/placeos-driver/driver-specs/runner.cr:415:3 in 'should_send'
  from drivers/nec/display/display_spec.cr:51:4 in '__crystal_main'
  from ../usr/share/crystal/src/crystal/main.cr:105:5 in 'main_user_code'
  from ../usr/share/crystal/src/crystal/main.cr:91:7 in 'main'
  from ../usr/share/crystal/src/crystal/main.cr:114:3 in 'main'
  from ???
... spec failed
... terminating driver gracefully
level=INFO time=2020-10-05T15:48:29Z source=driver.protocol message=IO terminated, exiting cleanly
Driver terminated with: 0
spec runner exited with 0